### PR TITLE
Add elasticsearch/ingest_pipeline metricset

### PR DESCRIPTION
**Pull request by joshdover**
## What does this PR do?

Depends on:
- [x] https://github.com/elastic/beats/pull/34273

Ports https://github.com/elastic/integrations/pull/4597 to metricbeat. This adds a new `ingest_pipeline` metricset to the `elasticsearch` module of Metricbeat. This module does the following:
- Fetches the Nodes Stats API's ingest metrics
- Ingest two types of documents:
	- Top-level pipeline metrics are ingested on every interval, including the counters for total time, documents processed, and failure counts. We also calculate a "self time" metric which will subtract the time spent on processor calls to other pipelines.
	- Processor-level metrics are ingested on a sampling rate (25% - every 4th interval by default, configurable). For these metrics, a separate document is created for every processor in every pipeline, on every ES node. For this reason, the sampling strategy is used to minimize the amount of data that is produced.
- Supports local and cluster scopes for better performance

The UI for visualizing this data will be included as a dashboard. Right now this is only being shipped in [an Agent integration](https://github.com/elastic/integrations/pull/4597) with support for Stack Monitoring and Metricbeat index patterns. In a follow up PR to Kibana, a link will be added from the Stack Monitoring UI to this dashboard, or direct the user to install the package to get the dashboard.

Here's a summary of the new fields that are added:
```yaml
- name: ingest_pipeline
  type: group
  release: beta
  description: Runtime metrics on ingest pipeline execution
  fields:
    - name: name
      type: wildcard
      description: Name / id of the ingest pipeline
    - name: total
      type: group
      description: Metrics on the total ingest pipeline execution, including all processors.
      fields:
        - name: count
          type: long
          description: Number of documents processed by this pipeline
        - name: failed
          type: long
          description: Number of documented failed to process by this pipeline
        - name: time.total.ms
          type: long
          description: Total time spent processing documents through this pipeline, inclusive of other pipelines called
        - name: time.self.ms
          type: long
          description: Time spent processing documents through this pipeline, exclusive of other pipelines called
    - name: processor
      type: group
      fields:
        - name: type
          type: keyword
          description: The type of ingest processor
        - name: type_tag
          type: keyword
          description: The type and the tag for this processor in the format "<type>:<tag>"
        - name: order_index
          type: long
          description: The order this processor appears in the pipeline definition
        - name: count
          type: long
          description: Number of documents processed by this processor
        - name: failed
          type: long
          description: Number of documented failed to process by this processor
        - name: time.total.ms
          type: long
          description: Total time spent processing documents through this processor
```

## Why is it important?

With the adoption of Agent integrations, ingest pipeline performance is very important to overall ingest performance. Users have very little insight into this today and giving them dashboards and metrics from the existing ES APIs is a great first start in improving the situation.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Add unit tests
- [ ] https://github.com/elastic/elasticsearch/pull/92950
- [x] Determine how we want to handle XPack mode. By default, xpack mode enables all metricset but we want to ship this as beta, so we may need to modify this behavior
	- Solved by https://github.com/elastic/beats/pull/34273

## Related issues

- Relates to https://github.com/elastic/kibana/issues/41936

### CODE REVIEW

File: metricbeat/docs/modules/elasticsearch/ingest.asciidoc [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+[[metricbeat-metricset-elasticsearch-ingest]]
+=== Elasticsearch ingest metricset
+
+beta[]
+
+include::../../../module/elasticsearch/ingest/_meta/docs.asciidoc[]
+
+This is a default metricset. If the host module is unconfigured, this metricset is enabled by default.
```
Comment (ruflin) [inline metricbeat/docs/modules/elasticsearch/ingest.asciidoc:14]: host module?

File: metricbeat/helper/elastic/elastic.go [hunk 2]
```diff
@@ -158,8 +158,31 @@ func NewModule(base *mb.BaseModule, xpackEnabledMetricsets []string, logger *log
 		return nil, errors.Wrapf(err, "could not unpack configuration for module %v", moduleName)
 	}
 
-	// These metricsets are exactly the ones required if xpack.enabled == true
-	raw["metricsets"] = xpackEnabledMetricsets
+	// Ensure all required metricsets are enabled when xpack.enabled == true
```
Comment (ruflin) [inline metricbeat/helper/elastic/elastic.go:161]: @miltonhultgren We should extend this later to add a flag that allows us to configure metricsets individually again even if xpack is enabled. This ensures we can have cluster stats with a longer cycle.

File: metricbeat/metricbeat.reference.yml [hunk 3]
```diff
@@ -275,6 +275,7 @@ metricbeat.modules:
     #- index
     #- index_recovery
     #- index_summary
+    #- ingest
```
Comment (ruflin) [inline metricbeat/metricbeat.reference.yml:278]: Slightly surprised by the name. Is this all of ingest or just ingest pipelines?

File: metricbeat/module/elasticsearch/ingest/_meta/data.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    },
+    "metricset":{
+        "host":"localhost",
+        "module":"elasticsearch",
+        "name":"ingest",
+        "rtt":44269
+    },
+    "elasticsearch":{
+        "ingest":{
+            "example": "ingest"
```
Comment (ruflin) [inline metricbeat/module/elasticsearch/ingest/_meta/data.json:15]: This does not look like a real example?

File: metricbeat/module/elasticsearch/ingest/data_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	mbtest "github.com/elastic/beats/v7/metricbeat/mb/testing"
+	"github.com/elastic/beats/v7/metricbeat/module/elasticsearch"
+)
+
+func createEsMuxer() *http.ServeMux {
+	mux := http.NewServeMux()
+	// mux.Handle("")
```
Comment (ruflin) [inline metricbeat/module/elasticsearch/ingest/data_test.go:35]: leftover?

File: metricbeat/helper/elastic/elastic.go [hunk 6]
```diff
@@ -158,8 +158,31 @@ func NewModule(base *mb.BaseModule, xpackEnabledMetricsets []string, logger *log
...
+	cfgdMetricsets, err := raw.GetValue("metricsets")
+	if err != nil || cfgdMetricsets == nil || len(cfgdMetricsets.([]interface{})) == 0 {
```
Comment (ruflin) [inline metricbeat/helper/elastic/elastic.go:163]: The code here and below does somehow look too complicate for what it does but currently not deep enough into the code to recommend adjustements.

**Review by ruflin — COMMENTED** [review]
Overall LGTM. Left some minor comment. +1 on moving forward with it.

File: metricbeat/helper/elastic/elastic.go (hunk 2 above)
Comment (miltonhultgren) [reply metricbeat/helper/elastic/elastic.go:161]: I think @joshdover has a PR that lays some foundation for that.

File: metricbeat/helper/elastic/elastic.go (hunk 6 above)
Comment (joshdover) [reply metricbeat/helper/elastic/elastic.go:163]: This is included in the blocking PR, let's discuss it here: https://github.com/elastic/beats/pull/34273/files#r1081193171

File: metricbeat/metricbeat.reference.yml (hunk 3 above)
Comment (joshdover) [reply metricbeat/metricbeat.reference.yml:278]: After discussion, I agree. Going to rename this to ingest pipelines.

File: metricbeat/helper/elastic/elastic.go (hunk 2 above)
Comment (joshdover) [reply metricbeat/helper/elastic/elastic.go:161]: Yes that PR should help without breaking the current functionality. Allows for enabling new metricsets without allow users to disable any required ones in xpack mode

File: metricbeat/docs/modules/elasticsearch/ingest.asciidoc (hunk 1 above)
Comment (joshdover) [reply metricbeat/docs/modules/elasticsearch/ingest.asciidoc]: Looks like this was from the metricset generator, it was present in some of the other docs for this module too so I fixed those as well.

**Comment by rdner** [comment]
@joshdover it's hard to tell why you have these changes locally are you sure everything is committed?

Any uncommitted files would show up in that error message.

Also, don't run `update` after `check`. `update` is supposed to change files which you need to review and commit them. Only after that, run `check` and confirm that no more changes are needed.

If you can't reproduce the issue on the CI, I recommend to try run `go clean -cache -modcache -i -r` locally before `make -C metricbeat check`. Sometimes when dependencies are not pinned to a tag, e.g `golang.org/x/exp v0.0.0-20220921023135-46d9e7742f1e`, it can cause problems.

File: metricbeat/module/elasticsearch/ingest_pipeline/_meta/data.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
+{
+  "@timestamp": "2023-01-19T19:08:28.770Z",
+  "elasticsearch": {
+    "cluster": {
+      "id": "XXXX",
```
Comment (klacabane) [inline metricbeat/module/elasticsearch/ingest_pipeline/_meta/data.json:5]: Curious if we have issues publishing cluster uuids and node ids ? They are exposed in the other metricsets

File: metricbeat/module/elasticsearch/ingest_pipeline/ingest_pipeline.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+		fetchCounter:           0,
+		sampleProcessorsEveryN: sampleProcessorsEveryN,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *IngestMetricSet) Fetch(report mb.ReporterV2) error {
+	shouldSkip, err := m.ShouldSkipFetch()
```
[the remark this answers is not in the document]
Comment (joshdover) [reply metricbeat/module/elasticsearch/ingest_pipeline/ingest_pipeline.go:97]: Good catch, thank you. I think I can just remove this entirely as I'm already setting the URI correctly on the lines below this. Do you agree?

File: metricbeat/module/elasticsearch/ingest_pipeline/_meta/data.json (hunk 7 above)
Comment (joshdover) [reply metricbeat/module/elasticsearch/ingest_pipeline/_meta/data.json:5]: I didn't want to publish them from my personal cluster I was using to test :) I can copy the ones from the other examples though.

File: metricbeat/module/elasticsearch/ingest_pipeline/ingest_pipeline.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (klacabane) [reply metricbeat/module/elasticsearch/ingest_pipeline/ingest_pipeline.go]: Sounds good

**Comment by amitkanfer** [comment]
🚢

**Pull request merged**